dhcpv4/client.go

```diff
@@ -165,14 +207,110 @@ func (c *Client) Exchange(ifname string, discover *DHCPv4, modifiers ...Modifier
 	conversation = append(conversation, request)
 
 	// Ack
-	ack, err := BroadcastSendReceive(sfd, rfd, request, c.ReadTimeout, c.WriteTimeout, MessageTypeAck)
+	ack, err := c.SendReceive(sfd, rfd, request, MessageTypeAck)
 	if err != nil {
 		return conversation, err
 	}
 	conversation = append(conversation, ack)
 	return conversation, nil
 }
 
+// SendReceive sends a packet (with some write timeout) and waits for a
+// response up to some read timeout value. If the message type is not
+// MessageTypeNone, it will wait for a specific message type
+func (c *Client) SendReceive(sendFd, recvFd int, packet *DHCPv4, messageType MessageType) (*DHCPv4, error) {
+	var laddr net.UDPAddr
+	if c.LocalAddr == nil {
+		laddr = net.UDPAddr{IP: net.IPv4bcast, Port: ServerPort}
+	} else {
+		if addr, ok := c.LocalAddr.(*net.UDPAddr); ok {
+			laddr = *addr
+		} else {
+			return nil, fmt.Errorf("Invalid local address: not a net.UDPAddr: %v", c.LocalAddr)
+		}
+	}
+	var raddr net.UDPAddr
+	if c.RemoteAddr == nil {
+		raddr = net.UDPAddr{IP: net.IPv4bcast, Port: ServerPort}
+	} else {
+		if addr, ok := c.RemoteAddr.(*net.UDPAddr); ok {
+			raddr = *addr
+		} else {
+			return nil, fmt.Errorf("Invalid remote address: not a net.UDPAddr: %v", c.RemoteAddr)
+		}
+	}
```

Comment:
How about simplifying this with:
`packetBytes, err := MakeRawPacket(packet.ToBytes(), c.RemoteAddr, c.LocalAddr)`

Reply:
Two reasons:
- we cannot pass the values directly (`net.Addr` vs `*net.UDPAddr`). We need the port in MakeRawPacket, and I'm kinda wary of moving the typecast there.
- default values are set up here if the struct has `nil` values for `RemoteAddr` and `LocalAddr` and we need those default values later in the same function (lines 253-254) anyway.

Reply:
How about passing a net.Addr to the MakeRawPacket function. In that way we won't be needing a type assertion and the default ports can be set while creating the raw packet as we are currently doing it.

Reply:
We'll still need to parse the IP address and convert it to a `net.IP` type to build the `ipv4.Header` instance is used there. 

I need to isolate the type checking in another function (to do the check on the broadcast address @insomniacslk asked below anyway) but I'd rather make `MakeRawPacket` private, so we don't have mismatch in the types exposed in the public interface. 

We could make the function take `srcIP, dstIP: net.IP, srcPort, dstPort int` instead of the current signature. I would rather have that function not "think" about what's in the address (I'm thinking right now that a `net.UnixAddr` can also IIRC be a `net.Addr`) and explicitly pass the parameters in.

Reply:
I'm for the `net.Addr` approach for the reasons that I've explained above (see comments around `MakeRawPacket`).
I generally agree with not making functions too smart, but in this case *some* function has to be smart. I think the code would be easier to follow if only the bottom functions do this, while the upper level functions maintain a higher level interface. The `net.UnixAddr` issue would be circumvented by the type assertion in the callee.
No strong opinions anyway, if you see problems with this approach let's discuss them here

Reply:
> and probably the function should renamed appropriately to be fair

The function deserves renaming - I didn't mention earlier to avoid getting into yak shaving, but since it came up let's do it. `MakeRawUDPPacket`?

> We are still going to get port and address in sendReceive

This is probably the strongest argument we have to do the type assertion here instead of `MakeRawPacket`, let's go for it
 
In summary:
* type assertion here, error if not `net.UDPAddr`
* pass source and destination by copy, not by pointer, as it doesn't impact performances (theoretically it should be even faster because we remove one level of indirection and use short jumps instead of long jumps, but really doesn't matter). Also no need to check for `nil`.